Fix parsing of /etc/redhat-releases

candrews opened the pull request:
#### Description:

Fix parsing of /etc/redhat-releases

#### Rationale:

The content of /etc/redhat-releases can be any of the following: Red Hat Enterprise Linux Server release 6.10 (Santiago) Red Hat Enterprise Linux Server release 7.9 (Maipo) Red Hat Enterprise Linux release 8.6 (Ootpa)
Red Hat Enterprise Linux release 8.5 (Ootpa)

The regex previously would only match values that did not have a release code name and did not have "Server", handling only values like this: Red Hat Enterprise Linux release 8.6

Meaning that the regex didn't match many /etc/redhat-releases, such as that in the ubi images:
registry.access.redhat.com/ubi7/ubi:latest
registry.access.redhat.com/ubi8/ubi:latest
registry.access.redhat.com/ubi9/ubi:latest

With this commit, ubi (and other Red Hat installations) are now matched by the regex resulting in correct detect as being Red Hat enabling rules to be applied appropriately.

#### Review Hints:

jan-cerny reviewed (changes requested):
Unfortunately it doesn't work for me with the command that you provided in the review hist (oscap-podman registry.access.redhat.com/ubi8/ubi:latest xccdf eval --profile xccdf_org.ssgproject.content_profile_stig) with a fresh data stream. I still get notapplicable for all the rules.

I think that there are multiple problems:

- you modified installed_OS_is_rhel7 but ubi8 is based  on rhel8
- you modifiied the check for a RHEVH (Red Hat Enterprise Virtualization Host) which isn't the system that runs in a container

You would have to modify shared/checks/oval/installed_OS_is_rhel8.xml but there the rpminfo test is used and it seems to be used correctly.

Could the problem be instead related to the RPM DB in Fedora might be incompatible with the RPM DB in UBI 8?

candrews commented:
> * you modified installed_OS_is_rhel7 but ubi8 is based  on rhel8

Yes, I've now done that and updated the PR. I'm sorry for that careless mistake.



> * you modifiied the check for a RHEVH (Red Hat Enterprise Virtualization Host) which isn't the system that runs in a container

Here's the test I've done:
`sudo oscap-podman registry.access.redhat.com/ubi8/ubi:latest xccdf eval --verbose DEVEL --verbose-log-file oscap.log --rule xccdf_org.ssgproject.content_rule_package_telnet-server_removed --profile xccdf_org.ssgproject.content_profile_stig /usr/share/xml/scap/ssg/content/ssg-rhel8-ds.xml`

I selected a rule arbitrarily to reduce the log size.

Here's the resulting log: 
[oscap.log](https://github.com/ComplianceAsCode/content/files/10942671/oscap.log)

The part that interested me is:
```
I: oscap: Evaluating definition 'oval:org.open-scap.cpe.rhel:def:8': Red Hat Enterprise Linux 8. [oscap(176546):oscap(7f2af785dd80):oval_resultDefinition.c:159:oval_result_definition_eval]
I: oscap:   Evaluating rpmverifyfile test 'oval:org.open-scap.cpe.rhel:tst:8': redhat-release is version 8. [oscap(176546):oscap(7f2af785dd80):oval_resultTest.c:1145:oval_result_test_eval]
I: oscap:     Querying rpmverifyfile object 'oval:org.open-scap.cpe.redhat-release:obj:3', flags: 0. [oscap(176546):oscap(7f2af785dd80):oval_probe.c:113:oval_probe_query_object]
I: oscap:     Creating new syschar for rpmverifyfile_object 'oval:org.open-scap.cpe.redhat-release:obj:3'. [oscap(176546):oscap(7f2af785dd80):oval_probe.c:136:oval_probe_query_object]
I: oscap:     Starting probe on URI 'queue://rpmverifyfile'. [oscap(176546):oscap(7f2af785dd80):oval_probe_ext.c:840:oval_probe_ext_handler]
D: oscap:     Sending message. [oscap(176546):oscap(7f2af785dd80):oval_probe_ext.c:486:oval_probe_comm]
D: oscap:     MSG -> SEXP [oscap(176546):oscap(7f2af785dd80):seap-packet.c:260:SEAP_packet_msg2sexp]
D: oscap: ("seap.msg" ":id" 0 (("rpmverifyfile_object" ":id" "oval:org.open-scap.cpe.redhat-release:obj:3" ":oval_version" "5.10.1" ) (("name" ":operation" 11 ":var_check" 1 ) "" ) (("epoch" ":operation" 11 ":var_check" 1 ) "" ) (("version" ":operation" 11 ":var_check" 1 ) "" ) (("release" ":operation" 11 ":var_check" 1 ) "" ) (("arch" ":operation" 11 ":var_check" 1 ) "" ) (("filepath" ":operation" 5 ":var_check" 1 ) "/etc/redhat-release" ) (("behaviors" ":nolinkto" "true" ":nomd5" "true" ":nosize" "true" ":nouser" "true" ":nogroup" "true" ":nomtime" "true" ":nomode" "true" ":nordev" "true" ":noconfigfiles" "true" ":noghostfiles" "true" ) ) ) ) [oscap(176546):oscap(7f2af785dd80):seap-packet.c:261:SEAP_packet_msg2sexp]
D: oscap:     packet size: 4624 [oscap(176546):oscap(7f2af785dd80):seap-packet.c:262:SEAP_packet_msg2sexp]
D: oscap:     Waiting for reply. [oscap(176546):oscap(7f2af785dd80):oval_probe_ext.c:545:oval_probe_comm]
D: oscap:     probe_common_main started [oscap(176546):common_main(7f2af48566c0):probe_main.c:192:probe_common_main]
D: oscap:     SEAP_add_probe [oscap(176546):common_main(7f2af48566c0):seap.c:163:SEAP_add_probe]
D: oscap:     icache worker ready [oscap(176546):icache_worker(7f2adffff6c0):icache.c:201:probe_icache_worker]
D: oscap:     Received packet [oscap(176546):input_handler(7f2adf7fe6c0):seap-packet.c:794:SEAP_packet_recv]
D: oscap: ("seap.msg" ":id" 0 (("rpmverifyfile_object" ":id" "oval:org.open-scap.cpe.redhat-release:obj:3" ":oval_version" "5.10.1" ) (("name" ":operation" 11 ":var_check" 1 ) "" ) (("epoch" ":operation" 11 ":var_check" 1 ) "" ) (("version" ":operation" 11 ":var_check" 1 ) "" ) (("release" ":operation" 11 ":var_check" 1 ) "" ) (("arch" ":operation" 11 ":var_check" 1 ) "" ) (("filepath" ":operation" 5 ":var_check" 1 ) "/etc/redhat-release" ) (("behaviors" ":nolinkto" "true" ":nomd5" "true" ":nosize" "true" ":nouser" "true" ":nogroup" "true" ":nomtime" "true" ":nomode" "true" ":nordev" "true" ":noconfigfiles" "true" ":noghostfiles" "true" ) ) ) ) [oscap(176546):input_handler(7f2adf7fe6c0):seap-packet.c:795:SEAP_packet_recv]
D: oscap:     packet size: 4624 [oscap(176546):input_handler(7f2adf7fe6c0):seap-packet.c:796:SEAP_packet_recv]
D: oscap:     probe_worker_runfn has started [oscap(176546):input_handler(7f2adeffd6c0):worker.c:81:probe_worker_runfn]
D: oscap:     handling SEAP message ID 0 [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:94:probe_worker_runfn]
I: oscap:     Switching probe to PROBE_OFFLINE_CHROOT mode. [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:1035:probe_worker]
I: oscap:     I will run rpmverifyfile_probe_main: [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:1113:probe_worker]
D: oscap:     omit verify attr: nolinkto [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nomd5 [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nosize [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nouser [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nogroup [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nomtime [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nomode [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: nordev [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: noconfigfiles [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     omit verify attr: noghostfiles [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:541:rpmverifyfile_probe_main]
D: oscap:     Collecting rpmverifyfile data, query: f="/etc/redhat-release" (5) [oscap(176546):probe_worker(7f2adeffd6c0):rpmverifyfile_probe.c:552:rpmverifyfile_probe_main]
D: oscap:     NOP [oscap(176546):probe_worker(7f2adeffd6c0):icache.c:434:probe_icache_nop]
D: oscap:     Signaling `notempty' [oscap(176546):probe_worker(7f2adeffd6c0):icache.c:459:probe_icache_nop]
D: oscap:     Waiting for icache worker to handle the NOP [oscap(176546):probe_worker(7f2adeffd6c0):icache.c:469:probe_icache_nop]
D: oscap:     Extracting item from the cache queue: cnt=1, beg=0 [oscap(176546):icache_worker(7f2adffff6c0):icache.c:220:probe_icache_worker]
D: oscap:     Signaling `notfull' [oscap(176546):icache_worker(7f2adffff6c0):icache.c:241:probe_icache_worker]
D: oscap:     Handling NOP [oscap(176546):icache_worker(7f2adffff6c0):icache.c:265:probe_icache_worker]
D: oscap:     Sync [oscap(176546):probe_worker(7f2adeffd6c0):icache.c:477:probe_icache_nop]
D: oscap:     old flag: 0, new flag: 4. [oscap(176546):probe_worker(7f2adeffd6c0):probe-api.c:688:probe_cobj_set_flag]
I: oscap:     Leaving chroot mode [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:1195:probe_worker]
D: oscap:     handler result = 0x7f2ac8000e60, return code = 0 [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:99:probe_worker_runfn]
D: oscap:     probe thread deleted [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:119:probe_worker_runfn]
D: oscap:     Sorting blocks & building iterator array [oscap(176546):probe_worker(7f2adeffd6c0):sexp-manip.c:1408:SEXP_list_sort]
D: oscap:     Iterator count = 0 [oscap(176546):probe_worker(7f2adeffd6c0):sexp-manip.c:1442:SEXP_list_sort]
D: oscap:     cnt = 0 [oscap(176546):probe_worker(7f2adeffd6c0):seap-message.c:138:SEAP_msgattr_exists]
D: oscap:     no-reply not set: sending full reply [oscap(176546):probe_worker(7f2adeffd6c0):seap.c:481:SEAP_reply]
D: oscap:     MSG -> SEXP [oscap(176546):probe_worker(7f2adeffd6c0):seap-packet.c:260:SEAP_packet_msg2sexp]
D: oscap: ("seap.msg" ":id" 0 ":reply-id" 0 (4 () () () ) ) [oscap(176546):probe_worker(7f2adeffd6c0):seap-packet.c:261:SEAP_packet_msg2sexp]
D: oscap:     packet size: 569 [oscap(176546):probe_worker(7f2adeffd6c0):seap-packet.c:262:SEAP_packet_msg2sexp]
D: oscap:     name=reply-id, value=0x7f2ac8040d80 [oscap(176546):probe_worker(7f2adeffd6c0):seap-message.c:71:SEAP_msg_free]
D: oscap:     probe_worker_runfn has finished [oscap(176546):probe_worker(7f2adeffd6c0):worker.c:179:probe_worker_runfn]
D: oscap:     Received packet [oscap(176546):oscap(7f2af785dd80):seap-packet.c:794:SEAP_packet_recv]
D: oscap: ("seap.msg" ":id" 0 ":reply-id" 0 (4 () () () ) ) [oscap(176546):oscap(7f2af785dd80):seap-packet.c:795:SEAP_packet_recv]
D: oscap:     packet size: 569 [oscap(176546):oscap(7f2af785dd80):seap-packet.c:796:SEAP_packet_recv]
D: oscap:     Message received. [oscap(176546):oscap(7f2af785dd80):oval_probe_ext.c:579:oval_probe_comm]
D: oscap:     name=(null), value=0x7f2ac8051e40 [oscap(176546):oscap(7f2af785dd80):seap-message.c:71:SEAP_msg_free]
I: oscap:     Test 'oval:org.open-scap.cpe.rhel:tst:8' requires that at least one object defined by 'oval:org.open-scap.cpe.redhat-release:obj:3' exists on the system. [oscap(176546):oscap(7f2af785dd80):oval_resultTest.c:906:_oval_result_test_evaluate_items]
I: oscap:     0 objects defined by 'oval:org.open-scap.cpe.redhat-release:obj:3' exist on the system. [oscap(176546):oscap(7f2af785dd80):oval_resultTest.c:918:_oval_result_test_evaluate_items]
I: oscap:     No item matching object 'oval:org.open-scap.cpe.redhat-release:obj:3' was found on the system. (flag=does not exist) [oscap(176546):oscap(7f2af785dd80):oval_resultTest.c:954:_oval_result_test_evaluate_items]
I: oscap:   Test 'oval:org.open-scap.cpe.rhel:tst:8' evaluated as false. [oscap(176546):oscap(7f2af785dd80):oval_resultTest.c:1164:oval_result_test_eval]
I: oscap: Definition 'oval:org.open-scap.cpe.rhel:def:8' evaluated as false. [oscap(176546):oscap(7f2af785dd80):oval_resultDefinition.c:170:oval_result_definition_eval]
```
That seems to point to the regexes I changed in this PR.

> Could the problem be instead related to the RPM DB in Fedora might be incompatible with the RPM DB in UBI 8?

I'm trying to think of ways to test this, do you have any suggestions?

candrews commented:
This PR doesn't fix https://github.com/OpenSCAP/openscap/issues/1942 (at least, not completely). But,  I still think it should be applied because the existing regex (as far as I can tell) would never match.

jan-cerny commented:
I'm sorry for the confusion. Let me clarify it.

The checks where you modify the regular expressions in this PR serve to check whether the installed system is Red Hat Enterprise Virtualization Host (RHEVH). Notice the `rhevh` in the check IDs. On these systems, the `/etc/redhat-release` file is provided by the `redhat-release-virtualization-host` RPM package and they have the following contents:

```
Red Hat Enterprise Linux release 8.6
```

and

```
Red Hat Enterprise Linux release 7.9
```

So the current check regex isn't broken, there is no `Server` in these files.

The confusion comes form that the same definitions also checks the `redhat-release` RPM package using the rpminfo test. That test is used for all normal variants of RHEL systems that aren't Red Hat Enterprise Virtualization Host systems.

Now I would like to investigate why this rpminfo test doesn't work in the oscap-podman environment or whether these 2 branches of the definitions can be consolidated.

candrews commented:
Thank you for these clarifications.

> Now I would like to investigate why this rpminfo test doesn't work in the oscap-podman environment

I think I figured that out: https://github.com/OpenSCAP/openscap/pull/1956

jan-cerny commented:
That's amazing!

jan-cerny commented:
I think that the changes related to RHEVH aren't needed, because the current regex matches the contents of the os-release files on RHEVH. The original intention of this PR was to fix the UBI 8 scan, but this turned out to be an issue with OpenSCAP and is fixed in https://github.com/OpenSCAP/openscap/pull/1956. Thanks for this! great job!
